### ISSUE TITLE
PerfDiff: absolute-budget strategies (P95 250ms, Mean 100ms) classify on the diff run alone — an unchanged over-budget benchmark permanently blocks CI

### ISSUE THREAD

**Issue by rjmurillo** [rjmurillo/moq.analyzers#1267]
## Summary

An internal code audit found that `RegressionStrategyHelper.FindRegressions` — the shared classifier behind `PercentileRegressionStrategy` (250 ms P95 budget) and `MeanWallClockRegressionStrategy` (100 ms mean budget) — classifies each benchmark by comparing **only the diff run's absolute metric** against a fixed threshold. The baseline value is never consulted:

1. **Permanent false block:** a benchmark whose absolute P95/mean legitimately sits above the budget is reported as a "regression" on **every** run — including runs where it is byte-for-byte identical to the baseline. Once any benchmark crosses the budget on `main`, every subsequent PR fails the perf gate until the budget or benchmark is changed.
2. **False pass (by these strategies):** a 2x slowdown that stays under the budget (e.g. 50 ms → 99 ms mean) is classified `Greater` — i.e. logged as *better than the threshold*.
3. Dead/odd code: the `Math.Abs(metric - thresholdValueNs) < 0.1D` branch is unreachable except when `metric == thresholdValueNs` exactly, and `ComparisonResult.Unknown` is reachable only for NaN.

This matters because PerfDiff's exit code is a required PR merge gate (`build/scripts/perf/ComparePerfResults.ps1` line 25 passes `--failOnRegression`; ADR-008). Failure mode 1 is a standing denial-of-merge; failure mode 2 quietly narrows what the budget strategies can catch. This is a specific correctness defect in the classification logic — distinct from #602 (externalizing the threshold *values*) and from the strategy test-coverage issues #596/#597.

## Current behavior

`src/tools/PerfDiff/BDN/Regression/RegressionStrategyHelper.cs:53-86`:

```csharp
    private static RegressionResult[] FindRegressions(
        BdnComparisonResult[] comparison,
        double thresholdValueNs,
        Func<Benchmark, double?> metricSelector)
    {
        List<RegressionResult> results = [];
        foreach (BdnComparisonResult result in comparison
            .Where(result => metricSelector(result.BaseResult) != null && metricSelector(result.DiffResult) != null))
        {
            double metric = metricSelector(result.DiffResult)!.Value;

            ComparisonResult conclusion;
            if (metric > thresholdValueNs)
            {
                conclusion = ComparisonResult.Lesser;
            }
            else if (metric < thresholdValueNs)
            {
                conclusion = ComparisonResult.Greater;
            }
            else if (Math.Abs(metric - thresholdValueNs) < 0.1D)
            {
                conclusion = ComparisonResult.Indistinguishable;
            }
            else
            {
                conclusion = ComparisonResult.Unknown;
            }

            results.Add(new RegressionResult(result.Id, result.BaseResult, result.DiffResult, conclusion));
        }

        return results.ToArray();
    }
```

Note line 62: `metricSelector(result.DiffResult)` is the only value read — `result.BaseResult`'s metric is filtered for null at line 60 and then never used. `HasRegression` (same file, lines 13–51) then puts every `Lesser` into the `worse` list and returns `worseCount > 0`, which `BenchmarkComparisonService.CompareAsync` (`src/tools/PerfDiff/BDN/BenchmarkComparisonService.cs:37-46`) turns into `regressionDetected = true` → `PerfDiff.CompareAsync` → exit 1 under `--failOnRegression`.

The two consumers (verified names and thresholds):

`src/tools/PerfDiff/BDN/Regression/PercentileRegressionStrategy.cs:15-22`:

```csharp
        RegressionMetricConfig config = new(
            DisplayThreshold: Threshold.Parse("250ms"),
            ThresholdValueNs: 250D * TimeUnitConstants.NanoSecondsToMilliseconds,
            MetricSelector: b => b.Statistics?.Percentiles?.P95,
            DisplayValueSelector: r => r.DiffResult.Statistics!.Percentiles!.P95 / TimeUnitConstants.NanoSecondsToMilliseconds,
            MetricName: "P95");
```

`src/tools/PerfDiff/BDN/Regression/MeanWallClockRegressionStrategy.cs:15-22`: identical shape with `Threshold.Parse("100ms")`, `100D * TimeUnitConstants.NanoSecondsToMilliseconds`, `b.Statistics?.Mean`, `MetricName: "Mean"`.

**Concrete input → wrong verdict/exit code:**

- Baseline P95 = 260 ms, diff P95 = 260 ms (no change at all): `2.6e8 ns > 2.5e8 ns` → `Lesser` → `"========== P95 regression detected …"` → **exit 1**, on this PR and every future PR. False block.
- Baseline mean = 50 ms, diff mean = 99 ms (~2x regression): `9.9e7 ns < 1.0e8 ns` → `Greater` → logged `"… better than the threshold …"`. Not flagged by either budget strategy (the Mann-Whitney ratio strategies may still catch it — see policy below).

## Desired behavior

Treat these two strategies as what their thresholds imply: **absolute budget checks that fire only for budget violations attributable to the change under test.** A benchmark is a regression for these strategies iff:

```text
diffMetric > budget                                  // the diff run violates the budget
AND (
    baseMetric <= budget                             // the violation is new (baseline was within budget), OR
    diffMetric > baseMetric * (1 + RelativeTolerance) // a pre-existing violation got meaningfully worse
)
```

with `RelativeTolerance = 0.05` (5%), matching the relative threshold the repo already uses in `P95RatioRegressionStrategy`/`MeanPercentageRegressionStrategy` (`Threshold.Parse("5%")`) — i.e. the codebase's own definition of "more than noise".

Defense of the predicate:

- *Unchanged over-budget benchmark*: `baseMetric > budget` and diff not >5% worse → **not** a regression (it becomes a logged pre-existing violation). Unblocks CI while keeping the condition visible in logs.
- *Crossing the budget*: baseline within budget, diff above → regression, exactly the budget check's purpose.
- *Over-budget and still degrading*: both above budget but diff >5% worse than baseline → regression. Without this arm, a benchmark could ratchet upward indefinitely once past the budget.
- *Under-budget regressions* (e.g. 50 ms → 99 ms): **deliberately not flagged by these strategies.** Relative regressions under budget are owned by the Mann-Whitney ratio strategies (`PercentageRegressionStrategy` 35% median ratio; `P95RatioRegressionStrategy` / `MeanPercentageRegressionStrategy` 5% + 0.5 ms). The budget strategies stop mislabeling such cases as "better" only in the sense that the pre-existing-violation bucket no longer pollutes the `better` list; a 50→99 ms case remains `Greater` here ("within budget") by design, and this policy is documented in code.

Also remove the dead `< 0.1D` branch and repurpose `Indistinguishable` for the pre-existing-violation bucket (excluded from both `better` and `worse` lists by the existing filtering in `HasRegression`), and `Unknown` for NaN metrics.

## Implementation plan

All changes in `src/tools/PerfDiff/BDN/Regression/RegressionStrategyHelper.cs`. The public surface (`HasRegression(BdnComparisonResult[], ILogger, RegressionMetricConfig, out RegressionDetectionResult)`) and both strategy classes are untouched.

1. **Replace `FindRegressions`** (complete member):

```csharp
    /// <summary>
    /// Tolerance applied when both the baseline and the diff run exceed the budget:
    /// the diff must be more than this fraction worse than the baseline to count as a
    /// regression. Matches the 5% relative threshold used by the ratio strategies.
    /// </summary>
    private const double RelativeTolerance = 0.05;

    private static RegressionResult[] FindRegressions(
        BdnComparisonResult[] comparison,
        double thresholdValueNs,
        Func<Benchmark, double?> metricSelector)
    {
        List<RegressionResult> results = [];
        foreach (BdnComparisonResult result in comparison
            .Where(result => metricSelector(result.BaseResult) != null && metricSelector(result.DiffResult) != null))
        {
            double baseMetric = metricSelector(result.BaseResult)!.Value;
            double diffMetric = metricSelector(result.DiffResult)!.Value;

            ComparisonResult conclusion;
            if (double.IsNaN(baseMetric) || double.IsNaN(diffMetric))
            {
                conclusion = ComparisonResult.Unknown;
            }
            else if (diffMetric <= thresholdValueNs)
            {
                // Within budget. Regressions that stay under the budget are owned by the
                // Mann-Whitney ratio strategies, not by this absolute budget check.
                conclusion = ComparisonResult.Greater;
            }
            else if (baseMetric <= thresholdValueNs || diffMetric > baseMetric * (1 + RelativeTolerance))
            {
                // Budget violation attributable to this change: either the baseline was
                // within budget (new violation), or a pre-existing violation got
                // meaningfully (>5%) worse.
                conclusion = ComparisonResult.Lesser;
            }
            else
            {
                // Pre-existing budget violation, unchanged (within tolerance) by this
                // change. Reported separately; must not block the gate.
                conclusion = ComparisonResult.Indistinguishable;
            }

            results.Add(new RegressionResult(result.Id, result.BaseResult, result.DiffResult, conclusion));
        }

        return results.ToArray();
    }
```

2. **Surface the pre-existing-violation bucket in `HasRegression`.** Insert the following block into `HasRegression` immediately before the line `details = new RegressionDetectionResult(config.MetricName, config.DisplayThreshold);` (currently line 49). This is an insertion only — the existing `better`/`worse` handling, including the geomean lines at lines 33 and 45, stays byte-for-byte as it is (a sibling issue on Inf/NaN ratio handling edits those geomean lines; keeping this change as a pure insertion avoids conflicts in either order):

```csharp
        List<RegressionResult> preExisting = notSame
            .Where(result => result.Conclusion == ComparisonResult.Indistinguishable)
            .OrderBy(k => k.Id, StringComparer.Ordinal)
            .ToList();
        foreach (RegressionResult preExistingResult in preExisting)
        {
            double value = config.DisplayValueSelector(preExistingResult);
            logger.LogWarning("test: '{TestId}' {MetricName} is {MetricValue:F3} ms, over the {Threshold} budget in both baseline and diff; pre-existing violation, not treated as a regression from this change", preExistingResult.Id, config.MetricName, value, config.DisplayThreshold);
        }
```

   (`logger.Log*` calls in this file already compile without `CA1848` pragmas — see the existing `LogInformation` calls in the same method.)

3. Do **not** modify `PercentageRegressionStrategy`, `P95RatioRegressionStrategy`, or `MeanPercentageRegressionStrategy` — the Mann-Whitney percentage strategies use `BenchmarkDotNetDiffer.FindRegressions` (a different method, `src/tools/PerfDiff/BDN/BenchmarkDotNetDiffer.cs:119`) and already compare baseline vs. diff statistically. Their Inf/NaN defects are a separate sibling issue.

4. **Test plumbing** (skip if already done by a sibling PerfDiff correctness issue): add the `ProjectReference` to `tests/PerfDiff.Tests/PerfDiff.Tests.csproj`:

```xml
  <ItemGroup>
    <ProjectReference Include="..\..\src\tools\PerfDiff\PerfDiff.csproj" />
  </ItemGroup>
```

   Everything tested here is public (`PercentileRegressionStrategy`, `MeanWallClockRegressionStrategy`, `RegressionStrategyHelper`, data contracts), so no `InternalsVisibleTo` is needed for this issue.

## Tests

Decision table implemented below (metric = P95 for `PercentileRegressionStrategy`, budget 250 ms = `2.5e8` ns; analogous single case for `MeanWallClockRegressionStrategy`, budget 100 ms = `1e8` ns):

| baseMetric | diffMetric | budget | expected verdict | rationale |
|---|---|---|---|---|
| 260 ms | 260 ms | 250 ms | **no regression** | unchanged pre-existing violation (was: permanent false block) |
| 200 ms | 300 ms | 250 ms | **regression** | violation newly introduced |
| 50 ms | 99 ms | 100 ms | **no regression (this strategy)** | under budget; 2x relative regressions are owned by the ratio strategies (`PercentageRegressionStrategy` et al.) — documented policy, not an oversight |
| 260 ms | 300 ms | 250 ms | **regression** | pre-existing violation made >5% worse |
| 260 ms | 265 ms | 250 ms | **no regression** | pre-existing violation within 5% tolerance |
| (null stats) | 300 ms | 250 ms | **no regression, no crash** | pair skipped by the null-metric filter |

New file `tests/PerfDiff.Tests/BudgetStrategyTests.cs` (no files/JSON needed — the budget path never touches `Measurements`). Uses the shared `ListLogger` from `tests/PerfDiff.Tests/ListLogger.cs` (defined in the EtlDiffer verdict issue; add it if not present):

```csharp
using Microsoft.Extensions.Logging;
using PerfDiff.BDN.DataContracts;
using PerfDiff.BDN.Regression;
using Xunit;

namespace PerfDiff.Tests;

public sealed class BudgetStrategyTests
{
    private const double Ms = 1_000_000D; // ns per ms (mirrors TimeUnitConstants.NanoSecondsToMilliseconds)

    private static Benchmark WithP95(double? p95Ns)
        => new()
        {
            FullName = "Suite.Bench1",
            Statistics = p95Ns is null ? null : new Statistics
            {
                Mean = p95Ns.Value,
                Median = p95Ns.Value,
                Percentiles = new Percentiles { P95 = p95Ns.Value },
            },
        };

    private static BdnComparisonResult[] Comparison(Benchmark baseline, Benchmark diff)
        => [new BdnComparisonResult("Suite.Bench1", baseline, diff)];

    [Theory]
    [InlineData(260, 260, false)] // unchanged pre-existing violation: must NOT block CI
    [InlineData(200, 300, true)]  // crossed the budget: regression
    [InlineData(260, 300, true)]  // pre-existing violation made >5% worse: regression
    [InlineData(260, 265, false)] // pre-existing violation within 5% tolerance: not a regression
    [InlineData(200, 240, false)] // within budget on both sides
    public void PercentileStrategy_IsBaselineAware(double baseMs, double diffMs, bool expectRegression)
    {
        ListLogger logger = new();
        BdnComparisonResult[] comparison = Comparison(WithP95(baseMs * Ms), WithP95(diffMs * Ms));

        bool hasRegression = new PercentileRegressionStrategy().HasRegression(comparison, logger, out _);

        Assert.Equal(expectRegression, hasRegression);
    }

    [Fact]
    public void PercentileStrategy_UnchangedOverBudget_LogsPreExistingViolationWarning()
    {
        ListLogger logger = new();
        BdnComparisonResult[] comparison = Comparison(WithP95(260 * Ms), WithP95(260 * Ms));

        bool hasRegression = new PercentileRegressionStrategy().HasRegression(comparison, logger, out _);

        Assert.False(hasRegression);
        Assert.Contains(logger.Entries, e => e.Level == LogLevel.Warning && e.Message.Contains("pre-existing", StringComparison.Ordinal));
    }

    [Fact]
    public void MeanWallClockStrategy_UnderBudgetTwoTimesRegression_NotFlaggedByThisStrategy()
    {
        // 50ms -> 99ms is a 2x regression but stays under the 100ms budget. The budget
        // strategy does not own this case; the ratio strategies (35% median ratio,
        // 5%+0.5ms mean/P95 ratio) are the detectors for relative regressions.
        ListLogger logger = new();
        Benchmark baseline = new() { FullName = "Suite.Bench1", Statistics = new Statistics { Mean = 50 * Ms, Median = 50 * Ms } };
        Benchmark diff = new() { FullName = "Suite.Bench1", Statistics = new Statistics { Mean = 99 * Ms, Median = 99 * Ms } };

        bool hasRegression = new MeanWallClockRegressionStrategy().HasRegression(Comparison(baseline, diff), logger, out _);

        Assert.False(hasRegression);
    }

    [Fact]
    public void PercentileStrategy_MissingBaselineStatistics_SkippedWithoutCrash()
    {
        ListLogger logger = new();
        BdnComparisonResult[] comparison = Comparison(WithP95(null), WithP95(300 * Ms));

        bool hasRegression = new PercentileRegressionStrategy().HasRegression(comparison, logger, out _);

        Assert.False(hasRegression);
    }
}
```

If #596 (mean-based strategy tests) and #597 (percentile/ratio strategy tests) are implemented, these tests belong in those suites — this table is exactly the "threshold conditions" coverage those issues call for.

## Constraints

- **No new package dependencies** (`Directory.Packages.props` pinned: Perfolizer 0.6.1, BenchmarkDotNet 0.15.8).
- Threshold **values** stay hard-coded exactly as today (250 ms / 100 ms / 5%); externalizing them is #602 and must not be entangled with this correctness fix. When #602 lands, `RelativeTolerance` should be externalized with the rest.
- Keep System.CommandLine 2.0.3 patterns untouched; no changes outside `RegressionStrategyHelper.cs` and the test project.
- Logging via existing `ILogger` only.
- PerfDiff targets `net8.0`; PerfDiff.Tests targets `net8.0`.

## Validation

1. `dotnet build` — 0 warnings, 0 errors.
2. `dotnet test --settings ./build/targets/tests/test.runsettings` — all tests pass, including the 4 existing `LoggingRegistrationTests`.
3. `dotnet format` after changes; commit any formatting fixes.

## Acceptance criteria

- [ ] A benchmark identical to its baseline never triggers a P95/Mean budget "regression", regardless of its absolute value.
- [ ] A benchmark crossing its budget (baseline within, diff above) triggers a regression → exit 1 under `--failOnRegression`.
- [ ] A pre-existing budget violation made >5% worse triggers a regression; within 5% it logs a named pre-existing-violation warning instead.
- [ ] The dead `Math.Abs(metric - thresholdValueNs) < 0.1D` branch is gone; `Unknown` is produced only for NaN metrics.
- [ ] Mann-Whitney ratio strategies (`PercentageRegressionStrategy`, `P95RatioRegressionStrategy`, `MeanPercentageRegressionStrategy`) are byte-for-byte unchanged.
- [ ] Decision-table tests above pass; build/test/format validation clean.

## Out of scope

- Externalizing threshold values and the new tolerance constant — #602.
- Inf/NaN median-ratio handling and geomean math in the ratio strategies and in `HasRegression`'s geomean lines — sibling correctness issue (this issue deliberately does not touch lines 33/45 of `RegressionStrategyHelper.cs`).
- Consolidating strategy duplication — #568.
- Making the budget strategies detect under-budget relative regressions — owned by the ratio strategies by design (documented in code by this change).

## Related issues

- #602 — externalize hard-coded thresholds (values used here stay as-is until then).
- #596 / #597 — strategy unit-test coverage (natural home for the decision-table tests).
- #568 — `IBenchmarkRegressionStrategy` duplication refactor (should rebase on this fix).
- ADR-008 — merge-gate contract; failure mode 1 is a standing false block of that gate.

**Comment by rjmurillo** [comment]
Fixed in #1309 (merged as 1340f475f519c638ced3453db52753b35028820b). Regression verdict logic corrected. Covered by new boundary tests in RatioRegressionStrategyTests.cs (58 PerfDiff tests total).
